tf-psa-crypto/drivers/builtin/include/mbedtls/ecp.h

```diff
@@ -345,6 +345,8 @@ mbedtls_ecp_group;
 
 #define MBEDTLS_ECP_MAX_BYTES    ((MBEDTLS_ECP_MAX_BITS + 7) / 8)
 #define MBEDTLS_ECP_MAX_PT_LEN   (2 * MBEDTLS_ECP_MAX_BYTES + 1)
+#define MBEDTLS_ECP_MAX_MPI_BYTES      ((MBEDTLS_ECP_MAX_BYTES + sizeof(mbedtls_mpi_uint)) & \
+                                        ~(sizeof(mbedtls_mpi_uint)-1))
```

Comment:
Needing this definition is a [code smell](https://en.wikipedia.org/wiki/Code_smell). The size of MPI limbs is an implementation detail that I would not expect `ecp.h` to leak. What are we doing that needs `MBEDTLS_ECP_MAX_MPI_BYTES` and not `MBEDTLS_ECP_MAX_BYTES`?

Reply:
Looking at the PR history, it seems to me that 8666b0fbc800bb2b1b200d156c0449d99dd39325 "Add MBEDTLS_ECP_MAX_MPI define" was a misguided attempt to fix a problem described in the commit message of a3ce63184955da2efc90e83f633bf79bd7aa78e2 "Refactor mbedtls_psa_generate_key_iop_complete()", but the change
```
-    *key_len = operation->ecp.d.n * sizeof(mbedtls_mpi_uint);
+    *key_len = PSA_BITS_TO_BYTES(operation->ecp.grp.nbits);
```
should be enough to do that.

Reply:
Note that I don't currently intend to do a full code review. But if I was doing a review, this would be a blocker — I want to understand what's going on here.

Reply:
> Looking at the PR history, it seems to me that https://github.com/Mbed-TLS/mbedtls/commit/8666b0fbc800bb2b1b200d156c0449d99dd39325 "Add MBEDTLS_ECP_MAX_MPI define" was a misguided attempt to fix a problem described in the commit message of https://github.com/Mbed-TLS/mbedtls/commit/a3ce63184955da2efc90e83f633bf79bd7aa78e2 "Refactor mbedtls_psa_generate_key_iop_complete()", but the change

[a3ce631](https://github.com/Mbed-TLS/mbedtls/commit/a3ce63184955da2efc90e83f633bf79bd7aa78e2 ) is just a refactor it brings checking on the size to the beginning of the function so if it fails I don't need to wast time generating a key this has nothing to do with [8666b0f](https://github.com/Mbed-TLS/mbedtls/commit/8666b0fbc800bb2b1b200d156c0449d99dd39325) which adds the new macro and replace the intermediate buffer with it I don't see any misguided attempts anywhere?

Reply:
What was the purpose of 8666b0fbc800bb2b1b200d156c0449d99dd39325? The commit message doesn't say: it says what is done but not why.

What went wrong before that commit? The code after a47b82c20a40ea486886f4b60ab699c5daf0d3a0 makes sense to me.

Reply:
> What was the purpose of https://github.com/Mbed-TLS/mbedtls/commit/8666b0fbc800bb2b1b200d156c0449d99dd39325?
the issue manifested itself with 521 but curve because the minimum size in bytes accommodate it is 66 bytes, However mbedtls_ecp_gen_privkey() function uses bignum and bignum grows to accommodate for the curve size but because the size of the limp is 8 bytes so the minimum to fit is 72 bytes which was higher than 66 so the size check failed so the buffer had to be big enough to fit the maximum bignum representation of curve size.

Reply:
I still don't see what does https://github.com/Mbed-TLS/mbedtls/commit/a3ce63184955da2efc90e83f633bf79bd7aa78e2 has to do with anything?

Reply:
After chatting, I now understand the original problem: in 30437e64083a78b1adc0b4245657703094f8ad74, `mbedtls_psa_generate_key_complete()` calls `memcpy` to copy data out of `operation->ecp.d.p`. This is wrong for another reason: it doesn't take endianness into account. Once the code is fixed to call `mbedtls_mpi_write_binary()`, that function takes care of fitting the output into the buffer. The output buffer must have the size expected by `psa_import_key`, i.e. `PSA_BITS_TO_BYTES(curve_bits)`. That aspect of the code is correct in a47b82c20a40ea486886f4b60ab699c5daf0d3a0.